File: dsa-problems/Trees/solutions/javascript/TRE-011-lab-bottom-view-shadow-limit.js
const readline = require("readline");

class Solution {
  bottomViewWithLimit(n, values, left, right, D) {
    if (n === 0) return [];

    const viewMap = new Map();
    const q = [{ u: 0, c: 0, d: 0 }];
    let minCol = 0;
    let maxCol = 0;

    while (q.length > 0) {
      const { u, c, d } = q.shift();

      viewMap.set(c, values[u]);
      
      if (c < minCol) minCol = c;
      if (c > maxCol) maxCol = c;

      if (d < D) {
        if (left[u] !== -1) q.push({ u: left[u], c: c - 1, d: d + 1 });
        if (right[u] !== -1) q.push({ u: right[u], c: c + 1, d: d + 1 });
      }
    }

    const result = [];
    for (let c = minCol; c <= maxCol; c++) {
      if (viewMap.has(c)) {
        result.push(viewMap.get(c));
      }
    }
    return result;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/)));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  const values = new Array(n);
  const left = new Array(n);
  const right = new Array(n);
  for (let i = 0; i < n; i++) {
    values[i] = parseInt(data[idx++], 10);
    left[i] = parseInt(data[idx++], 10);
    right[i] = parseInt(data[idx++], 10);
  }
  const D = idx < data.length ? parseInt(data[idx], 10) : 0;

  const solution = new Solution();
  const ans = solution.bottomViewWithLimit(n, values, left, right, D);
  console.log(ans.join(" "));
});

File: dsa-problems/Trees/solutions/python/TRE-011-lab-bottom-view-shadow-limit.py
import sys
sys.setrecursionlimit(200000)
from collections import deque

def bottom_view_with_limit(n: int, values: list[int], left: list[int], right: list[int], D: int) -> list[int]:
    if n == 0:
        return []
        
    view_map = {}
    q = deque([(0, 0, 0)]) # u, col, depth
    
    min_col, max_col = 0, 0
    
    # If root depth > D (impossible since root is 0 and D >= 0), but safe check
    if D < 0:
        return []
        
    while q:
        u, c, d = q.popleft()
        
        view_map[c] = values[u]
        min_col = min(min_col, c)
        max_col = max(max_col, c)
        
        if d < D:
            if left[u] != -1:
                q.append((left[u], c - 1, d + 1))
            if right[u] != -1:
                q.append((right[u], c + 1, d + 1))
                
    result = []
    for c in range(min_col, max_col + 1):
        if c in view_map:
            result.append(view_map[c])
            
    return result

def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    idx = 0
    n = int(data[idx]); idx += 1
    values = [0] * n
    left = [0] * n
    right = [0] * n
    for i in range(n):
        values[i] = int(data[idx]); idx += 1
        left[i] = int(data[idx]); idx += 1
        right[i] = int(data[idx]); idx += 1
    D = int(data[idx]) if idx < len(data) else 0
    ans = bottom_view_with_limit(n, values, left, right, D)
    sys.stdout.write(" ".join(str(x) for x in ans))

if __name__ == "__main__":
    main()
